File: 算法/排序/快速排序.cpp
//
// Created by Momo on 2023/7/1.
//

#include <iostream>

using namespace std;

void quick_sort(int* arr, int start, int end){
    if(start >= end) return;    // 划分的子数组大小为0或1，已经完成排序
    int i = start, j = end, pivot = arr[(start + end) / 2];
    while(i < j) {
        // 这两个 while 是分开进行的，如果第一个while找到了比 pivot 小的数
        // 就会进行第二个 while 找到了比 pivot 大的数
        while(arr[i] < pivot) i++;
        while(arr[j] > pivot) j--;
        // 说明还没有过完所有的数
        if(i <= j) {    // 这里是 = 处理重复的情况
            swap(arr[i], arr[j]);
            // 交换了以后防止这两个数再被遍历到，我们跳过他们
            i++; j--;
        }
    }

    quick_sort(arr, start, j);
    quick_sort(arr, i, end);
}


int main() {
    int a[] = {2,1,5,4,5,7,3, 6, 8} ;
    quick_sort(a, 0, (sizeof(a)/sizeof(int))-1);
    for (int i : a)
        cout << i << " ";
    cout << endl;
    return 0;
}

File: 算法/排序/快速排序.py
import random


# 快速排序
def quick_sort(arr: list):
    # 如果只有一个元素或空，说明就是排序好的
    if len(arr) <= 1:
        return arr
    # 获取基准值
    pivot = random.choice(arr)  # 每一次把第一个作为基准值的话，最坏情况调用栈会很长，最好选中间的
    # 比基准值小的元素数组
    less = [i for i in arr[1:] if i < pivot]
    greater = [i for i in arr[1:] if i >= pivot]
    # 递归调用此函数以加速分成小到大或大到一个步骤
    return quick_sort(less) + [pivot] + quick_sort(greater)


def quick_sort_2(arr: list, start: int, end: int):
    if start >= end:
        return
    left: int = start
    right: int = end
    pivot: int = arr[(start + end) // 2]

    while left < right:
        while arr[left] < pivot:
            left += 1
        while arr[right] > pivot:
            right -= 1

        if left <= right:
            arr[left], arr[right] = arr[right], arr[left]
            left += 1
            right -= 1
    quick_sort_2(arr, start, right)
    quick_sort_2(arr, left, end)


if __name__ == '__main__':
    ls = [4, 2, 1, 3, 8, 10, 3, 4]
    print(quick_sort(arr=ls))
